Compute feature in columns

m09 opened the pull request:
This is the first split of the refactoring. I tried to keep the changeset as minimal as possible in this PR to make it easy to review/merge.

Signed-off-by: Hugo Mougard <hugo@sourced.tech>

m09 commented:
Tests are passing locally and the error seems to be due to a bump in lookout.core.

m09 commented:
Rebased.

vmarkovtsev commented on lookout/style/format/features.py:
```diff
@@ -30,279 +66,389 @@ class FeatureId(Enum):
+MutableLayout = MutableMapping[FeatureGroup, List[T]]
+FeatureLayout = Layout[Mapping[FeatureId, T]]
+MutableFeatureLayout = MutableLayout[MutableMapping[FeatureId, T]]
+T_Feature = TypeVar("T_Feature", bound="Feature")
```
How about TVFeature? Funny, precise and does not break the naming conventions

vmarkovtsev commented on lookout/style/format/feature_extractor.py:
```diff
@@ -143,37 +109,48 @@ def _compute_feature_info(self) -> None:
+                        neighbour_index=node_index)
+                    selected_indices = []
+                    names = feature_pre_selection.names
+                    for index in range(len(names)):
```
enumerate?

vmarkovtsev commented on lookout/style/format/feature_extractor.py:
```diff
@@ -259,14 +241,14 @@ def extract_features(self, files: Iterable[File], lines: List[List[int]]=None) \
+                file_vnodes_iterator, file.path, index_labels=index_labels)
             if self.insert_noops:
-                file_vnodes = self._add_noops(list(file_vnodes), file.path,
+                file_vnodes = self._add_noops(list(file_vnodes_iterator), file.path,
```
`if self.insert_noops:` is still here... I recall that it is always True now, so what's the point?

m09 replied:
indeed, I removed it in the next PR but will back-port because it should be fixed ASAP

vmarkovtsev commented on lookout/style/format/features.py:
```diff
@@ -30,279 +66,389 @@ class FeatureId(Enum):
-                 ) -> Iterable[int]:
-        yield min(abs(sibling.start.line - node.end.line),
-                  abs(sibling.end.line - node.start.line))
+    def __init__(self, **kwargs: Any) -> None:
```
ctor not needed

vmarkovtsev commented on lookout/style/format/features.py:
```diff
@@ -30,279 +66,389 @@ class FeatureId(Enum):
-    def __call__(self, sibling: Union[VirtualNode, bblfsh.Node], node: VirtualNode
-                 ) -> Iterable[int]:
-        yield abs(sibling.start.offset - node.start.offset)
+    def __init__(self, **kwargs: Any) -> None:
```
ctor not needed

vmarkovtsev commented on lookout/style/format/features.py:
```diff
@@ -30,279 +66,389 @@ class FeatureId(Enum):
-    def __call__(self, sibling: Union[VirtualNode, bblfsh.Node], node: VirtualNode
-                 ) -> Iterable[int]:
-        yield abs(sibling.start.col - node.start.col)
+    def __init__(self, **kwargs: Any) -> None:
```
ctor not needed

vmarkovtsev commented on lookout/style/format/features.py:
```diff
@@ -30,279 +66,389 @@ class FeatureId(Enum):
-    def __init__(self, language: str) -> None:
-        RolesMixin.__init__(self, language)
-        self._internal_types_index = self._roles.INTERNAL_TYPES_INDEX
+    def __init__(self, **kwargs: Any) -> None:
```
ctor not needed

vmarkovtsev commented on lookout/style/format/features.py:
```diff
@@ -30,279 +66,389 @@ class FeatureId(Enum):
-    def __call__(self, sibling: Union[VirtualNode, bblfsh.Node], node: VirtualNode
-                 ) -> Iterable[int]:
-        yield sibling.y + 1 if sibling.y is not None else 0
+    def __init__(self, **kwargs: Any) -> None:
```
ctor not needed

vmarkovtsev commented on lookout/style/format/features.py:
```diff
@@ -30,279 +66,389 @@ class FeatureId(Enum):
-    def __init__(self, language: str) -> None:
-        TokensMixin.__init__(self, language)
-        self._reserved_index = self._tokens.RESERVED_INDEX
+    def __init__(self, **kwargs: Any) -> None:
```
ctor not needed

vmarkovtsev commented on lookout/style/format/features.py:
```diff
@@ -30,279 +66,389 @@ class FeatureId(Enum):
+class MultipleValuesFeature(Feature):
+    """Base type for features that produce multiple values."""
+
+    def __init__(self, **kwargs: Any) -> None:
```
ctor not needed - `__init__` is inherited

vmarkovtsev commented on lookout/style/format/features.py:
```diff
@@ -30,279 +66,389 @@ class FeatureId(Enum):
 
-    pass
+    def __init__(self, **kwargs: Any) -> None:
+        """Construct a CategoricalFeature."""
```
ctor not needed

vmarkovtsev commented on lookout/style/format/features.py:
```diff
@@ -30,279 +66,389 @@ class FeatureId(Enum):
 
-    pass
+    def __init__(self, **kwargs: Any) -> None:
+        """Construct a BagFeature."""
```
ctor not needed

vmarkovtsev commented on lookout/style/format/features.py:
```diff
@@ -30,279 +66,389 @@ class FeatureId(Enum):
-    def __call__(self, sibling: Union[VirtualNode, bblfsh.Node], node: VirtualNode
-                 ) -> Iterable[int]:
-        yield sibling.start.col
+    def __init__(self, **kwargs: Any) -> None:
```
ctor not needed

vmarkovtsev commented on lookout/style/format/features.py:
```diff
@@ -30,279 +66,389 @@ class FeatureId(Enum):
-    def __call__(self, sibling: Union[VirtualNode, bblfsh.Node], node: VirtualNode
-                 ) -> Iterable[int]:
-        yield sibling.start.line
+    def __init__(self, **kwargs: Any) -> None:
```
ctor not needed

vmarkovtsev commented on lookout/style/format/postprocess.py:
```diff
@@ -56,6 +56,7 @@ def filter_uast_breaking_preds(
     :param vnode_parents: `VirtualNode`-s' parents mapping as the LCA of the closest
                            left and right babelfish nodes.
     :param node_parents: Parents mapping of the input UASTs.
+    :param rule_winners: Winning rule for each sample.
```
numpy array of `Rule`-s (just to make sure)?

m09 replied:
numpy array of rule indices, I'll update the docstring since we can't give proper numpy.ndarray type hints :(

vmarkovtsev commented on lookout/style/format/virtual_node.py:
```diff
@@ -1,10 +1,11 @@
 from typing import Callable, Iterable, Mapping, NamedTuple, Optional, Set, Tuple, Union
 
-import bblfsh
+from bblfsh import Node
```
I would rather use `bblfsh.Node` everywhere to reduce the confusion between `AnyNode`, `VirtualNode` and `Node`

m09 replied:
I'll import as bblfsh_Node then if it's fine for you to stay consistent with the import style we prefer in this repo

vmarkovtsev commented on lookout/style/format/features.py:
```diff
@@ -30,279 +66,389 @@ class FeatureId(Enum):
+        super().__init__(**kwargs)
+
+    def _write_row(self, node: AnyNode, neighbour: AnyNode, output_row: numpy.ndarray) -> None:
+        output_row[0] = self._clip_int(abs(neighbour.start.col - node.start.col))
```
This will not work with `bblfsh.Node` - `start` attribute does not exist

vmarkovtsev commented on lookout/style/format/virtual_node.py:
```diff
@@ -144,3 +142,6 @@ def from_node(node: bblfsh.Node, file: str, path: str,
                               path=path)
+
+
+AnyNode = Union[VirtualNode, Node]
```
This is my concern... Interfaces of those two are very different, and our type hints are not always consistent.

Each Node used to be backed by a VirtualNode before. Is it possible to always feed VirtualNode-s and take `VirtualNode.node` attribute?

m09 replied:
parents never used to be backed by VirtualNode-s, that's why we have this union in features. I'll add a restriction on types for each feature to clarify everything

vmarkovtsev commented on lookout/style/format/features.py:
```diff
@@ -30,279 +66,389 @@ class FeatureId(Enum):
+        super().__init__(**kwargs)
+
+    def _write_row(self, node: AnyNode, neighbour: AnyNode, output_row: numpy.ndarray) -> None:
+        output_row[0] = self._clip_int(min(abs(neighbour.start.line - node.end.line),
```
Same here - `VirtualNode` is implied by using `start`

vmarkovtsev commented:
@m09 We are good to go once the tests pass!

zurk commented:
@vmarkovtsev let me review it too before merging. I will be fast.

m09 commented:
I changed lots of things so I think another round of review would be warranted @vmarkovtsev. Ready @zurk.

BTW the tests fail in typo while I changed nothing there, and only in 3.5.

vmarkovtsev commented on lookout/style/format/feature_extractor.py:
```diff
@@ -358,6 +344,84 @@ def composite_class_printables(self) -> List[str]:
+        for group in FeatureGroup:
+            neighbours[group] = []
+        neighbours[FeatureGroup.node].append([])
+        for _ in range(self.left_siblings_window):
```
```python
for group_id, size in zip([FeatureGroup.left, FeatureGroup.right, FeatureGroup.parents],
                          [self.left_siblings_window, self.right_siblings_window, self.parents_depth]):
    neighbours[group_id].extend([] for _ in range(size))
```

zurk commented on lookout/style/format/feature_extractor.py:
```diff
@@ -185,21 +160,26 @@ def populate_indices(feature_group: FeatureGroup, feature_names: Sequence[str],
+    def index_to_feature(self) -> Optional[IndexToFeature]:
         """Return the mapping from integer indices to the corresponding feature names."""
-        return self._index_to_feature
+        return self._index_to_feature if hasattr(self, "_index_to_feature") else None
```
I think it is better to raise an error in case there is no `_index_to_feature`. It is kind of unexpected situation.

zurk commented on lookout/style/format/feature_extractor.py:
```diff
@@ -185,21 +160,26 @@ def populate_indices(feature_group: FeatureGroup, feature_names: Sequence[str],
+    def feature_to_indices(self) -> Optional[FeatureLayout[Sequence[int]]]:
         """Return the mapping from feature names to the corresponding integer indices."""
-        return self._feature_to_indices
+        return self._feature_to_indices if hasattr(self, "_feature_to_indices") else None
```
same here

zurk commented on lookout/style/format/feature_extractor.py:
```diff
@@ -185,21 +160,26 @@ def populate_indices(feature_group: FeatureGroup, feature_names: Sequence[str],
+    @property
+    def features(self) -> Optional[FeatureLayout[Feature]]:
+        """Return the `Feature`-s used by this feature extractor."""
+        return self._features if hasattr(self, "_features") else None
```
and same here

zurk commented on lookout/style/format/tests/test_quality_report.py:
```diff
@@ -76,9 +76,11 @@ def __enter__(self):
+        self.extend(value.splitlines())
         del self._stringio    # free up some memory
         sys.stdout = self._stdout
+        print(value)
```
why do you change `__exit__()`?

m09 replied:
not printing the value on exit makes debugging very annoying when the test fails

zurk commented on lookout/style/format/visualizer/server.py:
```diff
@@ -2,21 +2,22 @@
 from lookout.style.format.virtual_node import VirtualNode
 
-logging.basicConfig(level="INFO")
+logging.basicConfig(level="DEBUG")
```
default one should be INFO I think.

m09 replied:
This is a debugging tool, default one should be DEBUG

zurk commented on lookout/style/format/visualizer/server.py:
```diff
@@ -46,33 +47,32 @@ def _vnode_to_dict(vnode: VirtualNode) -> Mapping[str, Any]:
 
-def _input_matrix_to_descriptions(X: numpy.ndarray, feature_extractor: FeatureExtractor
-                                  ) -> List[str]:
+Dict_or_str = Union[Dict[str, "Dict_or_str"], str]
```
what about naming?
`DictOrStr` maybe? it is type.

zurk commented on lookout/style/format/feature_extractor.py:
```diff
@@ -358,6 +344,84 @@ def composite_class_printables(self) -> List[str]:
+            if vnode.node:
+                closest_left_node_id = id(vnode.node)
+            if id(vnode) not in vnodes_y_set:
+                continue
```
it is kind of computation heavy.
we know that `vnode` and `vnode_y` have the same order, so we can compare with only one element from `vnode_y`
```python
 if vnode != vnode_y[vnode_y_index]:
    vnode_y_index += 1
    continue
```

m09 replied:
It's not really computation heavy though. Set creation is linear on the size of vnode_y and in operation is O(1) so both your version and this one are overall linear on the size of vnode_y. I think therefore we should decide mainly on readability and maintainability and I feel like the PR version is clearer and doesn't maintain an additional counter that can be moved/messed with that could introduce bug.

zurk commented on lookout/style/format/feature_extractor.py:
```diff
@@ -358,6 +344,84 @@ def composite_class_printables(self) -> List[str]:
+            if self.parents_depth:
+                parent = self._find_parent(i, vnodes, parents, closest_left_node_id)
+            else:
+                parent = None
```
`parent = self._find_parent(i, vnodes, parents, closest_left_node_id) if self.parents_depth else None`
?

zurk commented on lookout/style/format/features.py:
```diff
@@ -1,14 +1,61 @@
-from lookout.style.format.virtual_node import VirtualNode
+from lookout.style.format.virtual_node import AnyNode, VirtualNode
+
+FEATURES_NUMPY_TYPE = numpy.uint8
```
just `FEATURES_TYPE` ?

m09 replied:
It's more informative this way I feel.

m09 replied:
Please comment again if you think there is an upside to renaming!

zurk commented on lookout/style/format/features.py:
```diff
@@ -75,234 +202,290 @@ class BagFeature(MultipleValuesFeature):
+    """Base type for features that focus on a node among the neighbours of the current node."""
+
+    def __init__(self, neighbour_group: FeatureGroup, neighbour_index: int, **kwargs: Any) -> None:
+        """Construct an OrdinalFeature.
```
~OrdinalFeature~ FocusedFeature

zurk commented on lookout/style/format/features.py:
```diff
@@ -75,234 +202,290 @@ class BagFeature(MultipleValuesFeature):
+            raise RuntimeError("Cannot transform babelfish nodes into virtual nodes.")
+        return [None if node is None else node.node for node in nodes]
+
+    def _check_whitelist(self) -> None:
```
I can not find function usage :thinking:

zurk commented on lookout/style/format/features.py:
```diff
@@ -75,234 +202,290 @@ class BagFeature(MultipleValuesFeature):
+        self._names = [self.id.name]
+
+
+class FocusedFeature(Feature, Generic[TVAnyNode]):
```
maybe `NeighbourFeature`?

m09 commented on lookout/style/format/feature_extractor.py:
```diff
@@ -68,8 +36,8 @@ def __init__(self, *, language: str, left_siblings_window: int, right_siblings_w
                  no_labels_on_right: bool, select_features_number: Optional[int],
-                 remove_constant_features: bool, insert_noops: bool, debug_parsing: bool,
-                 return_sibling_indices: bool, selected_features: Optional[numpy.ndarray] = None,
+                 debug_parsing: bool, return_sibling_indices: bool,
```
I agree in general for the result of the fitting computations, here it's the result of the transform computation, we are not stored in those attributes in sklearn. (ie if we use extract_features more than once, the would get overriden).

m09 commented on lookout/style/format/feature_extractor.py:
```diff
@@ -143,37 +106,49 @@ def _compute_feature_info(self) -> None:
-                    for index in range(len(list(self._features[feature_name].names))):
+                    feature_id = FeatureId[feature_name]
+                    self._feature_to_indices[feature_group][node_index][feature_id] = []
+                    feature_class = FEATURE_CLASSES[FeatureId[feature_name]]
```
FEATURE_CLASSES is a constant mapping containing classes so it should stay uppercase and feature_class is a variable, not a class definition so it should stay snake-cased. Looks fine to me but feel free to comment again if I missed something.

m09 commented:
@zurk @vmarkovtsev thanks for the reviews! Ready for next round :running_man:

vmarkovtsev commented:
🎉
